### ISSUE TITLE
Rule Proposal: vue/no-arrow-for-props

### ISSUE THREAD

**Issue by nathanchase** [vuejs/eslint-plugin-vue#846]
**Please describe what the rule should do:**
The rule would enforce the use of using a return formatting style instead of an empty object with fat arrow formatting style for properties (`data`, `computed`, `methods`, etc.) inside components.

**What category should the rule belong to?**

- [X] Enforces code style

**Code examples that this rule should warn about:**

**BAD**
```vue
data: () => {
    return {
      value: this.getValue() // 'this' is not accessible
    };
},
```

**GOOD**
```vue
data() {
    return {
      value: this.getValue() // Now 'this' is accessible
    };
},
```

**Comment by nathanchase** [comment]
See this Codesandbox for an example of how using the arrow will result in not having access to `this`:
https://codesandbox.io/s/ly0mlmknx7

**Comment by manniL** [comment]
PS: the instance will be passed in as the first argument of the function

```js
data: (v) => {
    return {
      value: v.getValue()
    };
},
```

**Comment by rchl** [comment]
> PS: the instance will be passed in as the first argument of the function

Not the case with component methods though:

```js
methods: {
  getValue: (v) => {
     v.foo; // error
  },
},
```

**Comment by manniL** [comment]
@nathanchase that's correct. wouldn't make sense for methods :yum:

**Comment by mysticatea** [comment]
Is not [no-invalid-this](https://eslint.org/docs/rules/no-invalid-this) rule enough?

**Comment by nathanchase** [comment]
> Is not [no-invalid-this](https://eslint.org/docs/rules/no-invalid-this) rule enough?

Perhaps to solve the 'this' issue specifically, but additionally having a rule that defines how the top-level instance options (data, components, methods, head, etc.) are written to keep consistency across an entire project would be helpful? I'm not aware of any other way via Prettier or otherwise to achieve that consistency.

**Comment by SkyaTura** [comment]
I'm curious about which kind of a real use case that you would need `this` inside `data()`.
I always thought that is something to avoid. Could you show me an example?

**Comment by rchl** [comment]
> I'm curious about which kind of a real use case that you would need this inside data().

@SkyaTura You might want to create data based on passed props, for example.

But this rule should be more generic and apply to all component options (data, computed, methods, created, mounted, etc...). It makes even more sense in those places.

>  I want to force arrow functions on data and asyncData (I'm using Nuxt) on my project

Not sure why would you want to do that. The syntax is longer and 'messier' than function shorthand on object. And it will make it impossible to use `this`.

Using arrow function for `asyncData` wouldn't be a problem because that function is not bound to `this` (component doesn't exist yet) but IMO it's simpler and cleaner to use function shorthand.

Choosing between `asyncData() { ... }` and `asyncData = () => { ... }`, I'd choose former.
